### DISCUSSION TITLE
Only absolute URLs are supported

### DISCUSSION THREAD

**Discussion by Ihatetomatoes** [vercel/next.js#19013] [Help] [answered]
I have a question regarding API routes in Next.js.

I have a simple form with an email input to subscribe people to an email list.

Similar to @leerob Mailchimp integration but for Convertkit.
https://leerob.io/blog/mailchimp-next-js

Everything works fine locally but when I deploy the app to Vercel I get this error: `Only absolute URLs are supported`.

```javascript
import fetch from "isomorphic-unfetch";
import React from "react";
import { useForm } from "react-hook-form";

const SignupForm = () => {
  const { handleSubmit } = useForm();

  const onSubmit = async ({ email }) => {
    try {
      await fetch("/api/subscribe", {
        body: JSON.stringify({
          email,
        }),
        headers: {
          "Content-Type": "application/json",
        },
        method: "POST",
      });
    } catch (error) {
      console.log({ error });
      // Uh oh, something went wrong
    }
  };

  return (
    <>
      <form onSubmit={handleSubmit(onSubmit)}>
        ...
      </form>
    </>
  );
};

export default SignupForm;
```

I don't want to hardcode the full URL of the API route. 

Do I need to create an `env variable` for each environment?

Has anyone come across a similar issue and what is the solution?

**Comment by itaditya** [comment]
I don't have full context but this is what I noticed-

1. You are fetching inside an event handler that means it will always run on client side. This means you can use browser fetch which doesn't need absolute url.

2. Now Nextjs has inbuilt support of node-fetch so you can just assume that fetch global will be available everywhere. 

If you use fetch server side however like in getServerSideProps then you do need absolute url and its not an issue with Next.js but with node-fetch.

**Reply by Ihatetomatoes** [reply]
Thanks @itaditya, I believe I did try the fetch on a client but it gave was giving me the same error, that's why I have tried a few other methods like the  `isomorphic-unfetch`.

Someone suggested to [create system env variable](https://vercel.com/docs/build-step#system-environment-variables). I will give it a try today.

**Comment by Ihatetomatoes** [comment, marked as answer]
The issue was that I didn't have `env variable CONVERTKIT_API_URL` set in the preview part of the Vercel project settings. Adding it there solved my issues.

<img width="536" alt="img_env-variables-preview" src="https://user-images.githubusercontent.com/735672/98748602-21ca9a00-240e-11eb-9337-75aea9acf2ab.png">

**Reply by itaditya** [reply]
Awesome, how do you make those arrows though?

**Reply by Ihatetomatoes** [reply]
I use [Monosnap](https://monosnap.com/). Cool app that lets you create screenshots and annotate them.

**Reply by mvavassori** [reply]
I couldn't figure out the problem for 2 days. Thank you so much.

**Comment by itaditya** [comment]
Glad it worked. 

In your first code snippet there was nothing Convert Kit specific so I think that the env variable fixed it in browser side. But the issue might still remain server side.

**Reply by itaditya** [reply]
I think there is some convertkit wrapper code which prepends the env variable. 

But suppose you're using normal fetch with relative url then it will work on client side but not in Next server hooks like getStaticProps

**Comment by abhinav-anshul** [comment]
@itaditya  I tried everything from this thread, but nothing is really working out for me.

I have .env file with the variable : (notice the quotes)
`NEXT_PUBLIC_GRAPHCMS_ENDPOINT = 'https://api-eu-central-1.graphcms.com/v2/......secret.......'`

I'm using this in service/index.js folder :
`const BASE_URL: any = process.env.NEXT_PUBLIC_GRAPHCMS_ENDPOINT
`

When I'm trying to deploy to vercel:
`

> Build error occurred
--
20:08:59.462 | TypeError: Only absolute URLs are supported
20:08:59.463 | at getNodeRequestOptions (/vercel/path0/node_modules/cross-fetch/node_modules/node-fetch/lib/index.js:1327:9)

`

In the dashboard, I have this :(notice the quotes)



<img width="777" alt="Screenshot 2022-05-14 at 8 15 04 PM" src="https://user-images.githubusercontent.com/26833319/168432203-2e01dcae-239a-4611-b188-a907d91f4dc9.png">


I'm not sure what I'm doing wrong here, can someone please help me out

**Reply by mvavassori** [reply]
-Why are you using quotes and spaces?
-use a .env.local file
https://nextjs.org/docs/basic-features/environment-variables

**Reply by dryleaf** [reply]
Most of the times any quotes is the cause of this issue. This could also be accepted as a valid answer.
> -Why are you using quotes and spaces? -use a .env.local file https://nextjs.org/docs/basic-features/environment-variables
